applets: Add new power-profiles applet

vkareh opened the pull request:
This applet allows the user to set the active power profile (power-saver, balanced, performance). It's compatible with the power-profiles-daemon and tuned, which are the same DBUS interfaces that GNOME relies on for power profiles.


<img width="298" height="246" alt="Screenshot at 2025-09-03 15-48-49" src="https://github.com/user-attachments/assets/a9606d44-0ca8-4dc4-9e3c-73dc335c76f8" />
<img width="366" height="155" alt="Screenshot at 2025-09-03 15-50-14" src="https://github.com/user-attachments/assets/7a1e99a5-3a52-47a9-abab-4901be4ab774" />

Closes #201
Fixes #399

vkareh committed:
applets: Add new power-profiles applet

This applet allows the user to set the active power profile
(power-saver, balanced, performance). It's compatible with the
power-profiles-daemon and tuned, which are the same DBUS interfaces that
GNOME relies on for power profiles.

lukefromdc commented:
Are these power profiles different than the processor speed governors controlled by the Cpufreq applet?

vkareh commented:
> Are these power profiles different than the processor speed governors controlled by the Cpufreq applet?

Huh, good point. There is some overlap, but this relies on the UPower DBUS, rather than directly modifying the system file, and so does not require admin privileges to change the settings.

I'll do a bit of research to see whether it makes sense to update the CPUFreq applet instead and what that would mean.

Let's not merge yet.

vkareh commented:
Just got some time to get back to this. Here's a tl;dr:

- `cpufreq` gives you fine-grained CPU frequency and voltage control only; better for performance testing or benchmarking (or you have some specific CPU scaling needs)
- `power-profiles` gives you system-wide power management (CPU, GPU, disk, network, etc.); better for general laptop battery optimization

So yes, there's overlap on the CPU side and you probably _don't_ want to be using both together. I wonder if we can merge both and provide a "simple" vs "advanced/fine-tune" mode, or something like that.

lukefromdc reviewed (commented):
The applet basically works, though only in x11 and not the wayland session. This would be expected if any part of the core mate-power-manager process is called from the applet but could otherwise possibly be avoided. I tested both in the x11 session (worked) and in the wayland session with the platform check locally disabled (applet did not appear).

Note that with recent kernels we don't actually get a lot of power management/governor control on recent AMD procs, which can be a serious nuisance when say, running a full power desktop off a car battery and an inverter, engine off. Thus I wasn't able to see a lot of CPU frequency difference but that has noting to do with the applet but rather loss of access to the "user" governor in the kernel.

Also it is possible to get into trouble if both this and the CPUfreq applet are in-process in the panel: if you dismiss the dialog to ask for a passphrase and it pops back up, then try to change it from the new applet, then kill the panel to stop the dialog popping up, the panel will be caught in a crash/restart loop.  Only saw this once but this is something to be aware of as it can force a user to kill a session to get the panel back.

vkareh commented:
@lukefromdc I added a new separate commit that should allow the applet to show up in a Wayland session. Can you test this when you have a chance?

lukefromdc commented:
I just built this and got a finished build only to realize the applets were built out of process. We need to build in-process for wayland in order to be able to display the applets in the panel as we do not have xembed and thus do not have GtkPlug/GtkSocket to work with

vkareh committed:
applets: Add support for building applets in-process

This enables Wayland compatibility. Out-of-process applets use
GtkPlug/GtkSocket which is X11-specific and doesn't work in Wayland
sessions.

The default build remains out-of-process for backwards compatibility.

vkareh commented:
@lukefromdc okay, let's try this again :stuck_out_tongue: 

I just pushed a change to build in process. Use `-Dapplets-in-process=true` for meson or `--enable-applets-in-process` for make. I have no way of testing this in Wayland, but I can confirm they run in the mate-panel process in X11 when built with the new flag. The default is still out-of-process since I don't want anything to break accidentally...

lukefromdc commented:
The applet itself works in-process and in wayland, just needs to get "hooked up" to the backend and move the popup to the panel. The commit for in-process support in an of itself works. Doesn't crash the panel on addition or removal either, which are common new applet bugs

The applet builds fine in-process and the applet shows up in wayland, but at least in my setup it doesn't actually affect CPU frequency. Note that power profiles controlled by any means don't seem to do much on AMD with recent kernels at least with my kernel build options though.

Also the popup is in the center of the screen in wayland, I might have to fix that myself with GtkLayershell if it cannot be made a child window of the panel to position it. Same approach would be used as what I did for the volume control applet if it comes to this.

These are the runtime messages I got from `mate-panel --replace` with this in the panel under wayland, second message is from trying to change the frequency. Mate power manager not running as the main binary is x11-dependent:

```

(mate-panel:43753): ~/mate-power-manager.log-WARNING **: 12:38:04.968: not connected
Gdk-Message: 12:38:09.137: Window 0x557e134ef1f0 is a temporary window without parent, application will not be able to position it on screen.

(mate-panel:43753): Gdk-WARNING **: 12:38:09.137: Couldn't map as window 0x557e134ef1f0 as popup because it doesn't have a parent

(mate-panel:43753): ~/mate-power-manager.log-WARNING **: 12:38:09.137: not connected
Gdk-Message: 12:38:23.168: Window 0x557e134ff730 is a temporary window without parent, application will not be able to position it on screen.

(mate-panel:43753): Gdk-WARNING **: 12:38:23.168: Couldn't map as window 0x557e134ff730 as popup because it doesn't have a parent

(mate-panel:43753): ~/mate-power-manager.log-WARNING **: 12:38:23.168: not connected


```

vkareh committed:
power-profiles: Fix menu positioning in Wayland

Properly anchor the popup menu to the applet widget to fix the menu
appearing in the center of the screen under Wayland.

vkareh commented:
@lukefromdc I pushed a new commit that anchors the popup to the applet widget. I think this should solve the display issue. For the actual functionality of the applet you would need to use the [`power-profiles-daemon`](https://gitlab.freedesktop.org/upower/power-profiles-daemon) or [`tuned`](https://github.com/redhat-performance/tuned) (assuming these work on AMD).

lukefromdc reviewed (approved):
In wayland with the power manager (which is x11-only) not running, the applet has nothing to connect to but the menus now render in the correct place. We can worry about a wayland backend for this later, the existing CpuFreq applet does work in wayland though with current Debian I have not been able to make it passwordess for some reason.

For wayland it might be an option to use the same backends as the CpuFreq applet. Note that I don't have anything Intel and AMD stuff can be a little different with this stuff these days.
